server/src/main/java/org/opensearch/indices/replication/OngoingSegmentReplications.java

```diff
@@ -0,0 +1,195 @@
+
+/**
+ * Manages references to ongoing segrep events on a node.
+ *
+ * @opensearch.internal
+ */
+class OngoingSegmentReplications {
+
+    private final RecoverySettings recoverySettings;
+    private final IndicesService indicesService;
+    private final Map<ReplicationCheckpoint, CopyState> copyStateMap;
+    private final Map<DiscoveryNode, SegmentReplicationSourceHandler> nodesToHandlers;
+
+    /**
+     * Constructor.
+     * @param indicesService {@link IndicesService}
+     * @param recoverySettings {@link RecoverySettings}
+     */
+    OngoingSegmentReplications(IndicesService indicesService, RecoverySettings recoverySettings) {
+        this.indicesService = indicesService;
+        this.recoverySettings = recoverySettings;
+        this.copyStateMap = Collections.synchronizedMap(new HashMap<>());
+        this.nodesToHandlers = Collections.synchronizedMap(new HashMap<>());
+    }
+
+    /**
+     * Operations on the {@link #copyStateMap} member.
+     */
+
+    /**
+     * A synchronized method that checks {@link #copyStateMap} for the given {@link ReplicationCheckpoint} key
+     * and returns the cached value if one is present. If the key is not present, a {@link CopyState}
+     * object is constructed and stored in the map before being returned.
+     */
+    synchronized CopyState getCachedCopyState(ReplicationCheckpoint checkpoint) throws IOException {
+        if (isInCopyStateMap(checkpoint)) {
+            final CopyState copyState = fetchFromCopyStateMap(checkpoint);
+            // we incref the copyState for every replica that is using this checkpoint.
+            // decref will happen when copy completes.
+            copyState.incRef();
+            return copyState;
+        } else {
+            // From the checkpoint's shard ID, fetch the IndexShard
+            ShardId shardId = checkpoint.getShardId();
+            final IndexService indexService = indicesService.indexService(shardId.getIndex());
+            final IndexShard indexShard = indexService.getShard(shardId.id());
+            // build the CopyState object and cache it before returning
+            final CopyState copyState = new CopyState(checkpoint, indexShard);
+
+            /**
+             * Use the checkpoint from the request as the key in the map, rather than
+             * the checkpoint from the created CopyState. This maximizes cache hits
+             * if replication targets make a request with an older checkpoint.
+             * Replication targets are expected to fetch the checkpoint in the response
+             * CopyState to bring themselves up to date.
+             */
+            addToCopyStateMap(checkpoint, copyState);
+            return copyState;
+        }
+    }
+
+    void cancelReplication(DiscoveryNode node) {
+        if (nodesToHandlers.containsKey(node)) {
+            final SegmentReplicationSourceHandler handler = nodesToHandlers.remove(node);
+            handler.cancel("Cancel on node left");
```

Comment:
Since this method is not synchronized, please check the `handler` is not `null`: under races NPEs are guaranteed.